### PR TITLE
Build Lyra with Bazel 5.0.0

### LINKED ISSUE #76: Lyra fails to build in latest GitHub Actions virtual environment
In the `ubuntu-latest` and `macos-latest` virtual environments, the default Bazel version is `5.0.0`, which requires `build-tools 30.0.0` or newer version to work properly. Lyra is currently using `build-tools 29.0.3`, which will cause Lyra failed to build with the default Bazel in GitHub Actions virtual environments.

After my attempts, I successfully built Lyra with:

- `Bazel 5.0.0`
- `build-tools 30.0.3`
- `platforms android-30`
- `ndk 21.4.7075529`

Built successfully on hosts:

- `Ubuntu 20.04 (amd64)`
  - `linux-bin ✅`
  - `android-bin ✅`
  - `android-app ✅`
- `macOS 11 (Intel)`
  - `macos-bin ✅`
  - `android-bin ❌`
  - `android-app ❌`
- `macOS 12.2 (Apple Silicon)`
  - `macos-bin ✅`
  - `android-bin ❌`
  - `android-app ❌`

### PR DESCRIPTION
**Pull request by reekystive**
**This fixes #76.**

In the ubuntu-latest and macos-latest virtual environments, the default Bazel version is 5.0.0, which requires build-tools 30.0.0 or newer version to work properly. Lyra is currently using build-tools 29.0.3, which will cause Lyra failed to build with the default Bazel in GitHub Actions virtual environments.

This change will build Lyra with:

- `Bazel 5.0.0`
- `build-tools 30.0.3`
- `platforms android-30`
- `ndk 21.4.7075529` (unchanged)

Built successfully on hosts:

- `Ubuntu 20.04 (amd64)`
  - `linux-bin ✅`
  - `android-bin ✅`
  - `android-app ✅`
- `macOS 11 (Intel)`
  - `macos-bin ✅`
  - `android-bin ❌`
  - `android-app ❌`
- `macOS 12.2 (Apple Silicon)`
  - `macos-bin ✅`
  - `android-bin ❌`
  - `android-app ❌`

I also added actions to build on more platforms:

- `android-bin`
- `android-app`
- `linux-amd64`
- `macos-amd64`

GitHub Actions also uploads artifacts for download nightly builds now.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,16 +1,98 @@
...
+          name: lyra-android-bin
+          path: action-product
 
   android-app:
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - name: Build Android app
-        run: bazel build android_example:lyra_android_example --config=android_arm64 --copt=-DBENCHMARK
+      - name: Checkout repo
```
Comment (mchinen) [inline .github/workflows/ci.yml:34]: Thanks, it's useful to have these targets.  However, there are many steps that are duplicated across these.  Could you refactor as much as possible below this line with action composition?

https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/

**Comment by reekystive** [comment]
Commit cac8b94 in this PR also added artifacts upload job for:

- `android-bin`
- `android-app`
- `linux-amd64`
- `macos-amd64`

**Review by mchinen — COMMENTED** [review]
Thanks for the PR.  You may need to rebase as another PR touching ci.yml has just landed.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (reekystive) [reply .github/workflows/ci.yml:34]: Yes, I just saw that PR being merged, and I've already resolved the conflict now. This PR contains a change that is exactly equivalent to #15.

Comment (mchinen) [reply .github/workflows/ci.yml:34]: Great.  Please take a look at the refactoring comment.

Comment (mchinen) [reply .github/workflows/ci.yml:34]: Hi, just checking in on this.  Would you be able to address the refactoring comment?

Comment (mchinen) [reply .github/workflows/ci.yml:34]: Hi, I just want to note that I will be taking up this PR and will add the changes I mentioned myself.  If you prefer to do them yourself, please let me know.

Comment (reekystive) [reply .github/workflows/ci.yml:34]: I'm so sorry I'm a little busy at this time so I didn't pay attention to it. But I'm happy to do that if you haven't started yet!

**Comment by reekystive** [comment]
I've refactored the actions to reduce duplication, please let me know if there are other issues and I'll try to fix it.

**Comment by EwoutH** [comment]
@aluebs or @mchinen, any chance one of you could review (or merge) this PR?

**Comment by mchinen** [comment]
Sorry for overlooking this for so long.  Will merge shortly.

**Review by mchinen — APPROVED** [review]

**Comment by mchinen** [comment]
Hi @reekystive, after merging looks like this is failing due to the ANDROID_NDK_HOME path https://github.com/google/lyra/runs/8100486980?check_suite_focus=true#step:1:7.  Do you have an idea of how to resolve?

**Comment by reekystive** [comment]
It looks like the latest image doesn't have this version of NDK pre-installed.

https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#android

Let me try to update the NDK version used for building, or download an older version during CI.

**Comment by mchinen** [comment]
Thanks @reekystive.

**Pull request merged**